[2단계 - 로그인 구현하기] 엠제이(최민준) 미션 제출합니다.

theminjunchoi opened the pull request:
안녕하세요 머랭!
머랭은 머랭쿠키를 좋아하셔서 닉네임이 머랭이신가요?

이번 미션도 어찌저찌 요구사항은 만족하긴하는데 슬슬 코드양이 많아지면서 조금씩 더러워지는 것 같네요.

지난 pr에서 머랭 리뷰가 너무 맛있어서 이번에도 자유롭게, 머랭이 공부하면서 제 코드를 봤을 때 이해가 안가거나, 
개선점이 보이는 부분들을 자유롭게 리뷰해주시면 감사하겠습니다~!

theminjunchoi committed:
feat: sendRedirect 메서드 추가 및 로그인/리다이렉트 로직 구현

- `HttpResponse` 클래스에 `sendRedirect` 메서드 추가
  - 302 응답 상태 코드와 Location 헤더 설정
- `LoginServlet`에서 로그인 결과에 따라 `/index.html` 또는 `/401.html`로 리다이렉트 처리
  - 로그인 성공 시 `/index.html`로 이동
  - 로그인 실패 시 `/401.html`로 이동

theminjunchoi committed:
feat: LoginServlet에 POST 메서드 처리 로직 추가 및 로그인 실패 리다이렉트 개선

- POST 요청 시 handlePost 메서드로 분리하여 처리
- 계정 정보 누락 시 `/login` 리다이렉트 처리
- 405 Method Not Allowed 응답 처리 유지

theminjunchoi committed:
feat: RegisterServlet 추가 및 회원가입 처리 로직 구현

- GET 요청 시 register.html 페이지 반환
- POST 요청 처리: 회원가입 성공 시 `/`로 리다이렉트, 실패 시 `/register`로 리다이렉트
- InMemoryUserRepository 연동 및 회원 정보 저장 로직 추가
- Context 클래스에 `/register` 경로 RegisterServlet 등록
- 405 Method Not Allowed 응답 처리 로직 포함

theminjunchoi committed:
feat: HttpRequest POST 요청 본문 파싱 로직 추가 및 Http11Processor 헤더 처리 로직 구현

- HttpRequest 생성자에 POST 요청 본문(body) 파싱 로직 추가
- Http11Processor에서 HTTP 헤더 읽기 및 저장 로직 추가
- POST 요청 시 Content-Length 기반 본문 데이터 처리 로직 구현
- parseRequest 메서드 수정: 헤더 및 본문 파라미터 처리 지원

theminjunchoi committed:
feat: HttpRequest, HttpResponse에 쿠키 처리 추가

- `HttpCookie` 클래스 구현: 쿠키 파싱 및 관리 기능 제공
- `HttpRequest`에 쿠키 파싱 로직 추가: Cookie 헤더 기반 쿠키 정보 처리
- `HttpResponse`에 쿠키 추가/관리 기능 구현: Set-Cookie 헤더 지원
- `LoginServlet`에 JSESSIONID 쿠키 생성 및 처리 로직 추가

theminjunchoi committed:
feat: 세션 및 세션 관리 기능 추가

- `Session` 클래스 추가: 세션 ID, 속성 관리 및 무효화 처리
- `SessionManager` 싱글톤 클래스 구현: 세션 생성, 조회, 삭제 기능 제공
- `HttpRequest.getSession` 메서드 추가: JSESSIONID 기반 세션 처리 로직 구현
- `Manager` 인터페이스 수정: 커스텀 `Session` 타입으로 변경
- `LoginServlet` 수정: 로그인 상태 유지 및 세션에 사용자 정보 저장 로직 추가
- 로그인 상태 확인 시 세션 활용, 중복 로그인 상황 리다이렉션 처리

theminjunchoi committed:
refactor: HttpResponse 응답 생성 방식 개선

- StringBuilder 대신 Text Block 활용하여 응답 문자열 작성 로직 단순화
- 쿠키(Set-Cookie 헤더) 작성 로직을 별도의 StringBuilder로 분리하여 가독성 향상
- HTTP 상태 코드 및 메시지 포맷팅 방식 변경으로 코드 구조 간결화
- 기존 코드는 유지하며, 예외 처리 및 flush 로직 변경 없음

theminjunchoi commented:
@coderabbitai review

theminjunchoi committed:
feat: HttpMethod enum 추가

- HTTP 메서드를 관리하는 `HttpMethod` enum 추가: GET, POST, PUT, DELETE 등 지원
- `HttpRequest` 수정: 문자열 메서드 대신 `HttpMethod` enum 활용
  - 메서드 파싱 및 검증 로직 `HttpMethod.fromString`으로 일원화
  - `getMethod` 반환 타입을 `HttpMethod`로 변경
  - POST 요청 본문 처리 및 파라미터 로직에서 `HttpMethod.POST` 비교 사용
- 서블릿 클래스들(`RegisterServlet`, `LoginServlet`, `StaticResourceServlet`) 수정
  - 문자열 비교 대신 `HttpMethod` enum 기반으로 조건문 변경
  - 코드 가독성 및 유지보수성 개선
- 기타 코드 리팩터링: HttpRequest와 서블릿 간 HTTP 메서드 처리 방식 통일화

theminjunchoi committed:
feat: HttpStatus enum 추가

- HTTP 상태 관리를 위한 `HttpStatus` enum 추가
  - 상태 코드 및 Reason Phrase 관리: OK, NOT_FOUND, INTERNAL_SERVER_ERROR 등
  - 코드 기반 상태 조회 기능: `fromCode` 메서드로 매핑 처리
- `HttpResponse`에서 HTTP 상태 관리 로직을 `HttpStatus` 기반으로 변경
  - 기존 상태 코드 사용 로직 제거 및 enum 활용하도록 수정
  - `setStatus` 메서드 오버로딩: 숫자 코드 및 `HttpStatus` 수용
  - 상태 코드와 Reason Phrase 출력 방식 간소화
- 서블릿에서 상태 코드 설정 시 `HttpStatus` 활용
  - `RegisterServlet`, `LoginServlet`, `StaticResourceServlet`에 적용
  - 상태 코드 상수 대신 `HttpStatus` 참조로 가독성 향상
- `StaticFileResult` 및 정적 리소스 처리 로직 리팩터링
  - 상태 코드를 `HttpStatus` enum으로 변경
  - 상태 코드 관리 일원화로 코드 통일 및 유지보수성 개선

theminjunchoi committed:
refactor: 서블릿 초기화 로직 Context에서 Tomcat으로 이동

- Context에서 ServletContainer 초기화 로직 제거 및 Tomcat으로 이동
- Tomcat 생성자에서 ServletContainer 직접 생성 및 설정
- Context 간소화: ServletContainer 관련 코드 제거
- Tomcat 클래스에 서블릿 등록 로직 추가
  - `/login`, `/register`, `/*` 경로에 각각 `LoginServlet`, `RegisterServlet`, `StaticResourceServlet` 등록
- 코드 가독성 및 역할 분리 개선

theminjunchoi committed:
refactor: Context 클래스 삭제

- ServletContainer 생성 로직 제거 후 Context 클래스 삭제
- 역할 분리를 통해 Tomcat 클래스에 관련 로직 통합 및 유지보수성 개선

theminjunchoi committed:
fix: RegisterServlet 오류 페이지 HTML 태그 수정

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -25,42 +29,69 @@ public void init() {
         }
 
-        response.setStatus(405);
+        if (HttpMethod.POST == request.getMethod()) {
```
Enum 간의 `==` 비교는 메서드 call stack 오버헤드 제거로 인한 성능 최적화를 고려하신 건가요??
만약 그런 거라면 좋은데요?? 멋진 코드네요 ㅎㅎ

theminjunchoi replied:
사실 거기까지는 생각하지 못했는데요..ㅎㅎ

단순 컴파일 타임에 타입 미스매치를 잡아주고 NPE를 발생시키지 않기 위해 `==`을 사용했습니다!
머랭이 말씀하신 것처럼 미세한 call-stack 오버헤드도 제거할 수 있겠네요!

cookie-meringue replied:
아하 ㅎㅎ 알겠습니다!👍

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -25,42 +29,69 @@ public void init() {
 
     @Override
     public void service(final HttpRequest request, final HttpResponse response) {
-        if ("GET".equals(request.getMethod())) {
+        if (HttpMethod.GET == request.getMethod()) {
             handleGet(request, response);
             return;
         }
 
-        response.setStatus(405);
+        if (HttpMethod.POST == request.getMethod()) {
+            handlePost(request, response);
+            return;
+        }
+
+        response.setStatus(HttpStatus.METHOD_NOT_ALLOWED);
         response.write("<html><body><h1>405 Method Not Allowed</h1></body></html>");
     }
```
Servlet 간 `service()` 내부 로직이 전부 동일하고, 앞으로도 그럴 것 같다고 생각이 듭니다!

Interface - 구현체 구조를 Abstract Class 에서 Template Method 패턴을 사용해 handleGet()과 handlePost()만 자식 클래스에서 구현하도록 하는 건 어떻게 생각하시는지 궁금합니다!

```java
public abstract class Servlet{

    // init(), destroy() 등
    // ....

    public final void service(final HttpRequest request, final HttpResponse response) {
        if (HttpMethod.GET == request.getMethod()) {
            handleGet(request, response);
            return;
        }

        if (HttpMethod.POST == request.getMethod()) {
            handlePost(request, response);
            return;
        }

        response.setStatus(HttpStatus.METHOD_NOT_ALLOWED);
        response.write("<html><body><h1>405 Method Not Allowed</h1></body></html>");
    }

    protected abstract void handleGet(final HttpRequest request, final HttpResponse response);

    protected abstract void handlePost(final HttpRequest request, final HttpResponse response);
}

public class LoginServlet implements Servlet {

    // init(), destroy() 등
    // ....

    @Override
    protected void handleGet(final HttpRequest request, final HttpResponse response) {
        // 로그인 GET 관련 로직
    }

    @Override
    protected void handlePost(final HttpRequest request, final HttpResponse response) {
        // 로그인 POST 관련 로직
    }
}
```


이러면 각 Servlet들에서 GET, POST 를 분류해서 메서드를 호출하거나 중복 코드를 상위로 분리해서 유지보수성을 챙길 수 있고, response 에 405 관련 응답을 작성하는 로직도 묶을 수 있을 것 같습니다.
`createErrorPage()` 와 같은 Servlet 의 공통 메서드도 Servlet 으로 올릴 수 있겠네요!

상속과 템플릿 메서드 패턴 자체의 단점도 있으니 고려해보시고 의견 말씀해주시면 감사하겠습니다!

theminjunchoi replied:
엇 너무 좋은 피드백인데요,
이 부분은 step3에 할거라서 다음 리뷰때 반영해보겠습니다!!

이미 객체분리가 어느정도 돼서 step3에서 할 게 없더라고요..!

cookie-meringue replied:
앗 ㅎㅎ 알겠습니다!

cookie-meringue commented on tomcat/src/main/java/org/apache/coyote/http11/HttpRequest.java:
```diff
@@ -65,4 +83,33 @@ public String getParameter(final String name) {
+        return cookies.getValue(name);
+    }
+
+    public Session getSession() {
+        return getSession(true);
+    }
+
+    public Session getSession(final boolean create) {
+        final String sessionId = getCookieValue("JSESSIONID");
+        
+        if (sessionId != null) {
+            final Session session = SessionManager.getInstance().findSession(sessionId);
+            if (session != null) {
+                return session;
+            }
+        }
+        
+        if (create) {
+            return SessionManager.getInstance().createSession();
+        }
+        
+        return null;
+    }
```
Request 가 Session 을 생성하는 책임을 가지는 것이 조금 어색하게 느껴지는데 엠제이는 어떻게 생각하시나요??

theminjunchoi replied:
엇 저도 어색한 것 같습니다.
SessionManager에서 Session을 생성하도록 수정해볼게요!

cookie-meringue replied:
넵 ㅎㅎ 알겠습니다!

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -25,42 +29,69 @@ public void init() {
             log.info("로그인 성공: 회원 조회 결과 - {}", user);
+
+            // 세션에 사용자 정보 저장
+            final Session session = request.getSession(true);
+            session.setAttribute("user", user);
+            response.addCookie("JSESSIONID", session.getId());
+            response.sendRedirect("/");
```
Session 에 user 객체를 저장하도록 구현하신 이유는 무엇인가요??
세션이 살아 있는 동안 유저의 정보가 변경된다면 데이터 불일치가 발생할 것 같은데, 엠제이의 의견이 궁금합니다!

theminjunchoi replied:
오 그럴 수 있겠네요!
user가 아니라 userId를 갖고 있게 하면 좋을 것 같은데
이것도 step3에서 반영해볼게요!

cookie-meringue replied:
네 알겠습니다 ㅎㅎ

cookie-meringue commented on tomcat/src/main/java/com/techcourse/servlet/LoginServlet.java:
```diff
@@ -25,42 +29,69 @@ public void init() {
+        response.sendRedirect("/401.html");
+    }
+
+    private User getUser(final Session session) {
+        return (User) session.getAttribute("user");
     }
```
Session 에 다음과 제네릭 메서드를 추가해서 Class 클래스를 활용하는 것은 어떨까요??

```java
public <T> T getAttribute(String name, Class<T> type) {
        Object value = values.get(name);

        if (value == null) {
            return null;
        }
        
        if (type.isInstance(value)) {
            return type.cast(value);
        }

        throw new ClassCastException("...");
    }
```

theminjunchoi replied:
오 좋은데요!?
반영해볼게요!

cookie-meringue replied:
넵!!

cookie-meringue commented on tomcat/src/main/java/org/apache/catalina/SessionManager.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Map;
+import java.util.UUID;
+
+public class SessionManager implements Manager {
```
Singleton 클래스가 final 이 아니라면, 상속을 통한 다중 인스턴스 생성이 가능합니다!

> 중요한 부분은 아닌 것 같네요.. 넘어가셔도 좋습니다!

theminjunchoi replied:
오 그럴 수 있겠어요!
final 추가해볼게요!

cookie-meringue replied:
넵!!

cookie-meringue commented on tomcat/src/main/java/org/apache/coyote/http11/HttpCookie.java:
```diff
@@ -0,0 +1,51 @@
+            return cookieMap;
+        }
+        
+        final String[] cookiePairs = cookieHeader.split(";");
+        for (final String cookiePair : cookiePairs) {
+            final String[] keyValue = cookiePair.trim().split("=", 2);
+            if (keyValue.length == 2) {
+                cookieMap.put(keyValue[0].trim(), keyValue[1].trim());
+            }
+        }
```
이건 반영 안하셔도 되는데, "="을 기준으로 split 하면 base64로 인코딩된 값을 제대로 불러올 수 없는 문제가 있습니다!

네이버 메인 페이지에서 개발자 도구로 찾은 쿠키 예시 첨부드립니다!

key: `NID_JKL`
value: `CTDGKjpvss+Dq8goxib9JuWMO87LeZjASUNcriu7AeM=`

<img width="648" height="174" alt="image" src="https://github.com/user-attachments/assets/dc586a9a-9af8-4ce2-9638-6f047368c94e" />


Base 64 말고도 "="이 들어가는 값이라면 제대로 동작하지 않을 것 같습니다!

theminjunchoi replied:
엇 맞아요
저도 머랭이 발견해주신 것과 같은 경우를 고민을 했었는데요!

```java
final String[] keyValue = cookiePair.trim().split("=", 2);
```

기존코드처럼 해주면 limit가 2로 걸려서 
```
[0]: "NID_JKL"
[1]: "CTDGKjpvss+Dq8goxib9JuWMO87LeZjASUNcriu7AeM="
```
이런식으로 파싱이 잘 될 것 같아요..!

cookie-meringue replied:
헉 제가 두 번째 매개변수를 보지 못했네요 ㅎㅎ.. 알겠습니다!

cookie-meringue reviewed (changes requested):
고생 많으셨습니다!
큰 수정 사항이 없는 것 같으니, Comment만 달아주시면 머지하도록 하겠습니다!

> 머랭은 제 이름의 앞글자를 따서 지었습니다 ㅋㅋㅋ
> 대현 -> 머현 -> 머랭

theminjunchoi committed:
test: ThreadPoolsTest와 SynchronizationTest 테스트 코드 수정

- ThreadPoolsTest에서 기대 값 수정 및 테스트 통과하도록 변경
  - FixedThreadPool: PoolSize 2, QueueSize 1로 수정
  - CachedThreadPool: PoolSize 3으로 수정
- SynchronizationTest에서 메서드 동기화 추가
  - calculate, getSum, setSum 메서드에 synchronized 키워드 적용
- 코드 가독성 및 동작 안정성 개선

theminjunchoi committed:
refactor: getAttribute type 지정하도록 수정

theminjunchoi reviewed (commented):
안녕하세요 머랭!
꼼꼼한 리뷰 감사합니다~!
반영하고 다시 요청 드릴게요!

theminjunchoi committed:
refactor: SessionManager final 추가

theminjunchoi committed:
refactor: Session 생성책임 SessionManager로 위임

theminjunchoi committed:
refactor: 불필요한 주석 제거

cookie-meringue reviewed (approved):
고생 많으셨습니다!
